### PR TITLE
Subscription ids cache

### PR DESCRIPTION
**Pull request by dankraw**
Introducing a subscription id cache, based on [Znode czxid](https://zookeeper.apache.org/doc/r3.5.5/zookeeperProgrammers.html#sc_zkDataModel_znodes).
Subscription ids will be used for subscription identification in the underlying storage  instead of plain strings as  it is currently.

### CODE REVIEW

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCache.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import pl.allegro.tech.hermes.domain.notifications.InternalNotificationsBus;
+import pl.allegro.tech.hermes.domain.notifications.SubscriptionCallback;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class NotificationAwareSubscriptionIdsCache implements SubscriptionIds, SubscriptionCallback {
+
+    private final Map<SubscriptionId, SubscriptionName> ids = new HashMap<>();
```
Comment (jewertow) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCache.java]: Don't we have to care for thread safety?

File: hermes-consumers/src/test/groovy/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCacheTest.groovy [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        def sub2 = subscriptionNames.next()
+        def sub1Id = SubscriptionId.of(sub1)
+        def sub2Id = SubscriptionId.of(sub2)
+        subscriptionCache.listActiveSubscriptionNames() >> []
+
+        when:
+        subscriptionIds.start()
+
+        then:
+        !subscriptionIds.getSubscriptionName(sub1Id).ifPresent()
```
Comment (jewertow) [inline hermes-consumers/src/test/groovy/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCacheTest.groovy]: `ifPresent()`? I suspect you wanted to use `.isPresent()`.

Comment (dankraw) [reply hermes-consumers/src/test/groovy/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCacheTest.groovy]: ouch, good catch.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCache.java (hunk 1 above)
Comment (dankraw) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCache.java]: Hmm it supposed to be used by a single thread, but as long as it's not used anywhere yet, I think it should allow concurrent access. Good point, I'll fix it, thanks. :)

File: hermes-consumers/src/test/groovy/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCacheTest.groovy (hunk 2 above)
Comment (dankraw) [reply hermes-consumers/src/test/groovy/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCacheTest.groovy]: fixed.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCache.java (hunk 1 above)
Comment (dankraw) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/NotificationAwareSubscriptionIdsCache.java]: Fixed.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/ZookeeperSubscriptionIdProvider.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        this.curatorFramework = curatorFramework;
+        this.zookeeperPaths = zookeeperPaths;
+    }
+
+    @Override
+    public SubscriptionId getSubscriptionId(SubscriptionName name) {
+        return Optional.ofNullable(getZnodeStat(name))
+                .map(Stat::getCzxid)
+                .map(czxid -> SubscriptionId.from(name, czxid))
+                .orElseThrow(() -> new IllegalStateException(
```
Comment (jewertow) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/ZookeeperSubscriptionIdProvider.java]: I'm not sure if it's appropriate use of `IllegalStateException`. 
Documentation says: "Signals that a method has been invoked at an illegal or inappropriate time. In other words, the Java environment or Java application is not in an appropriate state for the requested operation."
Maybe could we throw dedicated exception?

Comment (dankraw) [reply hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/subscription/id/ZookeeperSubscriptionIdProvider.java]: IMHO it fits actually - when there is no znode related to specific subscription, it means the method was called in inappropriate time, probably after the subscription was removed (or was not created yet, but it's impossible).
I will rethink your suggestion, thanks.

**Pull request merged**